src/astral/data-utils.js

```diff
@@ -0,0 +1,86 @@
+const tokenCache = {
+    access: null,
+    apiKey: null,
+    refresh: null,
+    expires: null
+};
+
+const _getTokenFromDb = () => new Promise((resolve, reject) => {
+    const req = indexedDB.open("firebaseLocalStorageDb", 1);
+    console.log("Created db req", req)
+    req.onsuccess = (event) => {
+        const db = req.result;
+        const trans = db.transaction("firebaseLocalStorage", IDBTransaction.READ)
+        const objStore = trans.objectStore("firebaseLocalStorage");
+
+        const cursorReq = objStore.openCursor();
+
+        cursorReq.onsuccess = (event) => {
+            console.log("Create cursor req", cursorReq.result)
+            resolve(cursorReq.result.value.value.stsTokenManager);
+        };
+    }
+
+    req.onerror = reject;
+});
+
+const getAccessToken = async () => {
+    if (!tokenCache.expires) {
+        const tokenData = await _getTokenFromDb();
+
+        tokenCache.access = tokenData.accessToken;
+        tokenCache.apiKey = tokenData.apiKey;
+        tokenCache.expires = tokenData.expirationTime;
+        tokenCache.refresh = tokenData.refreshToken;
+    } else if (tokenCache.expires < Date.now()) {
```

Comment:
Consider it expired 1 minute before real expiration. Otherwise, if it will expire in 100ms, but it takes more than 100ms to send the request to the server (either based on latency or on clock skew between local and remote server clocks), then the request will fail.